quinn-proto/src/connection/paths.rs

```diff
@@ -163,3 +170,453 @@ impl RttEstimator {
+        }
+
+        let next_mtu = (self.search_lower_bound as i32 + self.search_upper_bound as i32) / 2;
+
+        // Binary search stopping condition
+        if ((next_mtu - self.probing_udp_payload_size as i32).unsigned_abs() as u16)
+            < BINARY_SEARCH_MINIMUM_CHANGE
+        {
+            // Special case: if the upper bound is far enough, we want to probe it as a last
+            // step (otherwise we will never achieve the upper bound)
+            if self
+                .search_upper_bound
+                .saturating_sub(self.probing_udp_payload_size)
+                >= BINARY_SEARCH_MINIMUM_CHANGE
+            {
+                return Some(self.search_upper_bound);
+            }
+
+            return None;
+        }
+
+        Some(next_mtu as u16)
+    }
+
+    /// Notifies the [`MtuDiscovery`] that a packet has been ACKed
+    pub fn acked(&mut self, space: SpaceId, packet_number: u64, packet_bytes: u16) {
+        // MTU probes are only sent in application data space
+        if space != SpaceId::Data {
+            return;
+        }
+
+        self.ack_probe(packet_number);
+
+        // Reset the black hole counter if a packet the size of the current MTU or larger
+        // has been acknowledged
+        if packet_bytes >= self.current_mtu
+            && self
+                .largest_acked_mtu_sized_packet
+                .map_or(true, |pn| packet_number > pn)
+        {
+            self.black_hole_counter = 0;
+            self.largest_acked_mtu_sized_packet = Some(packet_number);
+        }
+    }
+
+    /// Updates `current_mtu` if the packet number corresponds to the in-flight MTU probe
+    fn ack_probe(&mut self, packet_number: u64) {
+        if Some(packet_number) == self.probe_packet_number {
+            self.probe_packet_number = None;
+            self.lost_probes = 0;
+            self.current_mtu = self.probing_udp_payload_size;
+            trace!(max_udp_payload_size = self.current_mtu, "new MTU detected");
+        }
+    }
+
+    /// Returns the packet number of the in-flight MTU probe, if any
+    pub fn in_flight_mtu_probe(&self) -> Option<u64> {
+        self.probe_packet_number
+    }
+
+    /// Notifies the [`MtuDiscovery`] that the in-flight MTU probe was lost
+    pub fn lost_probe(&mut self) {
+        self.probe_packet_number = None;
+        self.lost_probes += 1;
+    }
+
+    /// Notifies the [`MtuDiscovery`] that a non-probe packet was lost
+    pub fn lost_non_probe_packet(&mut self, now: Instant, packet_number: u64, packet_size: u16) {
+        // Keep track of lost packets higher than the base size, and reset the current MTU
+        // if they reach the threshold
+        if packet_size > BASE_UDP_PAYLOAD_SIZE
+            && self
+                .largest_acked_mtu_sized_packet
+                .map_or(true, |pn| packet_number > pn)
+        {
+            self.black_hole_counter += 1;
+            if self.black_hole_counter > BLACK_HOLE_THRESHOLD {
```

Comment:
Should this threshold be configurable?

Reply:
It would be easy to add, but it might make the API less easy to change in a backwards-compatible way (e.g. what if we want to use a different black hole detection mechanism in the future?)

Putting my API designer's hat on, I'd wait until the current approach has proven itself _and_ users ask for more control

Reply:
SGTM.